perception/traffic_light_recognition/traffic_light_classifier/src/nodelet.cpp

```diff
@@ -18,55 +18,58 @@
-  image_transport_.reset(new image_transport::ImageTransport(pnh_));
-  pnh_.param<bool>("approximate_sync", is_approximate_sync_, false);
+  using std::placeholders::_1;
+  using std::placeholders::_2;
+  is_approximate_sync_ = this->declare_parameter("approximate_sync", false);
   if (is_approximate_sync_) {
     approximate_sync_.reset(new ApproximateSync(ApproximateSyncPolicy(10), image_sub_, roi_sub_));
     approximate_sync_->registerCallback(
-      boost::bind(&TrafficLightClassifierNodelet::imageRoiCallback, this, _1, _2));
+      std::bind(&TrafficLightClassifierNodelet::imageRoiCallback, this, _1, _2));
   } else {
     sync_.reset(new Sync(SyncPolicy(10), image_sub_, roi_sub_));
     sync_->registerCallback(
-      boost::bind(&TrafficLightClassifierNodelet::imageRoiCallback, this, _1, _2));
+      std::bind(&TrafficLightClassifierNodelet::imageRoiCallback, this, _1, _2));
   }
-  ros::SubscriberStatusCallback connect_cb = boost::bind(&TrafficLightClassifierNodelet::connectCb, this);
-  std::lock_guard<std::mutex> lock(connect_mutex_);
-  tl_states_pub_ = pnh_.advertise<autoware_perception_msgs::TrafficLightStateArray>(
-    "output/traffic_light_states", 1, connect_cb, connect_cb);
 
-  int classifier_type;
-  pnh_.param<int>(
-    "classifier_type", classifier_type, TrafficLightClassifierNodelet::ClassifierType::HSVFilter);
+  image_sub_.subscribe(this, "input/image", "raw", rclcpp::QoS{1}.get_rmw_qos_profile());
+  roi_sub_.subscribe(this, "input/rois", rclcpp::QoS{1}.get_rmw_qos_profile());
+
+  tl_states_pub_ = this->create_publisher<autoware_perception_msgs::msg::TrafficLightStateArray>(
+    "output/traffic_light_states", rclcpp::QoS{1});
+
+  int classifier_type = this->declare_parameter(
+    "classifier_type", static_cast<int>(TrafficLightClassifierNodelet::ClassifierType::HSVFilter));
   if (classifier_type == TrafficLightClassifierNodelet::ClassifierType::HSVFilter) {
-    classifier_ptr_ = std::make_shared<ColorClassifier>(nh_, pnh_);
+    classifier_ptr_ = std::make_shared<ColorClassifier>(this);
   } else if (classifier_type == TrafficLightClassifierNodelet::ClassifierType::CNN) {
 #if ENABLE_GPU
-    classifier_ptr_ = std::make_shared<CNNClassifier>(nh_, pnh_);
+    classifier_ptr_ = std::make_shared<CNNClassifier>(this);
 #else
-    NODELET_ERROR("please install CUDA, CUDNN and TensorRT to use cnn classifier");
+    RCLCPP_ERROR(
+      this->get_logger(), "please install CUDA, CUDNN and TensorRT to use cnn classifier");
 #endif
   }
 }
 
-void TrafficLightClassifierNodelet::connectCb()
-{
-  std::lock_guard<std::mutex> lock(connect_mutex_);
-  if (tl_states_pub_.getNumSubscribers() == 0) {
-    image_sub_.unsubscribe();
-    roi_sub_.unsubscribe();
-  } else if (!image_sub_.getSubscriber()) {
-    image_sub_.subscribe(*image_transport_, "input/image", 1);
-    roi_sub_.subscribe(pnh_, "input/rois", 1);
-  }
-}
+// TODO(mitsudome-r): port this to ROS2 once unsubscribe() is available in ROS2
```

Comment:
It's probably not as easy as setting the Subscriber to `nullptr`? That worked for a regular subscription at least.

Reply:
I see. I will try out and investigate a bit more.

Reply:
Looks like they had unsubscribe function for the `image_transport::SubscriberFilter` and  `message_filters::Subscriber`